File: src/main/java/network/crypta/launcher/LauncherLog.java
package network.crypta.launcher;

import java.io.PrintWriter;
import java.io.StringWriter;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.Locale;

/**
 * Minimal logging helpers for the launcher package.
 *
 * <p>- `logDebug()` prints only when `CRYPTA_LAUNCHER_DEBUG` is truthy.
 *
 * <p>- `logWarn()` and `logError()` always print to stderr with a timestamped prefix.
 */
public final class LauncherLog {
  private enum Lvl {
    DEBUG,
    INFO,
    WARN,
    ERROR
  }

  private static volatile Boolean debugEnabled;

  private LauncherLog() {}

  public static void logDebug(String msg) {
    emit(Lvl.DEBUG, msg, null);
  }

  public static void logDebug(String msg, Throwable t) {
    emit(Lvl.DEBUG, msg, t);
  }

  public static void logWarn(String msg) {
    emit(Lvl.WARN, msg, null);
  }

  public static void logWarn(String msg, Throwable t) {
    emit(Lvl.WARN, msg, t);
  }

  public static void logError(String msg) {
    emit(Lvl.ERROR, msg, null);
  }

  public static void logError(String msg, Throwable t) {
    emit(Lvl.ERROR, msg, t);
  }

  private static boolean isDebugEnabled() {
    Boolean cached = debugEnabled;
    if (cached != null) {
      return cached;
    }
    synchronized (LauncherLog.class) {
      if (debugEnabled == null) {
        String v = System.getenv("CRYPTA_LAUNCHER_DEBUG");
        if (v != null) {
          v = v.toLowerCase(Locale.ROOT).trim();
        }
        debugEnabled =
            "1".equals(v) || "true".equals(v) || "yes".equals(v) || "y".equals(v) || "on".equals(v);
      }
      return debugEnabled;
    }
  }

  private static String tsNow() {
    return LocalDateTime.now().format(DateTimeFormatter.ISO_LOCAL_TIME);
  }

  private static void emit(Lvl level, String msg, Throwable t) {
    if (level == Lvl.DEBUG && !isDebugEnabled()) {
      return;
    }
    String base = tsNow() + " [Launcher/" + level.name() + "] " + msg;
    if (t == null) {
      System.err.println(base);
      return;
    }
    StringWriter sw = new StringWriter();
    t.printStackTrace(new PrintWriter(sw));
    System.err.println(base + "\n" + sw);
  }
}

File: src/main/kotlin/network/crypta/launcher/LauncherLog.kt
package network.crypta.launcher

import java.io.PrintWriter
import java.io.StringWriter

/**
 * Minimal logging helpers for the launcher package.
 * - `logDebug()` prints only when the environment variable `CRYPTA_LAUNCHER_DEBUG` is set to a
 *   truthy value ("1", "true", "yes").
 * - `logWarn()` and `logError()` always print to stderr with a timestamped prefix.
 *
 * We keep this lightweight to avoid adding a logging framework dependency in the launcher.
 */
private enum class Lvl {
  DEBUG,
  @Suppress("unused") INFO,
  WARN,
  ERROR,
}

private val debugEnabled: Boolean by lazy {
  val v = System.getenv("CRYPTA_LAUNCHER_DEBUG")?.lowercase()?.trim()
  v == "1" || v == "true" || v == "yes" || v == "y" || v == "on"
}

private fun tsNow(): String =
  java.time.LocalDateTime.now().format(java.time.format.DateTimeFormatter.ISO_LOCAL_TIME)

private fun emit(level: Lvl, msg: String, t: Throwable? = null) {
  if (level == Lvl.DEBUG && !debugEnabled) return
  val base = "${tsNow()} [Launcher/${level.name}] ${msg}"
  if (t == null) {
    System.err.println(base)
  } else {
    val sw = StringWriter()
    t.printStackTrace(PrintWriter(sw))
    System.err.println(base + "\n" + sw.toString())
  }
}

fun logDebug(msg: String, t: Throwable? = null) = emit(Lvl.DEBUG, msg, t)

fun logWarn(msg: String, t: Throwable? = null) = emit(Lvl.WARN, msg, t)

@Suppress("unused") fun logError(msg: String, t: Throwable? = null) = emit(Lvl.ERROR, msg, t)
